Add basic session cookie handling

N00bySumairu opened the pull request:
With these two commits basic session cookie handling seems to work with what little testing I've been able to do so far (I'll test it some more tomorrow).

One potential caveat: With the current implementation, if the setting to continue the previous session is not enabled, then even if you stay in the app new connections to other rooms will not use a previous session cookie. This might not be desirable for #1. A possible fix would be clearing the stored session cookie on app launch instead of not passing a previous session cookie to new connections if there is one.

N00bySumairu committed:
Add cookie settings and settings helper class

N00bySumairu committed:
Implement basic session cookie handling

When a establishing a new connection ConnectionManager will pass on a
previous session cookie if the continue_prev_session setting is turned on.
The new connection will use this session cookie in the initial handshake.

A connection that receives a Set-Cookie header with a session cookie
will forward this cookie to the ConnectionManager,
which in turn will store this cookie for future use.

N00bySumairu committed:
Use the nick from a previous session if available

If a snapshot-event includes a nick from a previous session
the websocket client will now correctly emit a NickChangeEvent

CylonicRaider committed:
Add final blank lines

What's more amazing than the IDE autoformatter's configurability? The
configuration knobs it *lacks*.

CylonicRaider commented:
Hmm... The `updateSessionCookie` methods are curious. They look like event handler methods (raised in response to receiving a `Set-Cookie` header), and are indeed called by someone different from the caller of the other methods. It would seem more appropriate to move (and appropriately rename) them into `ConnectionListener`, or a new `IdentityListener`, with `EuphoriaWebSocketClient` using the regular `submitEvent` to launch a new `SessionCookieEvent` OSLT on its way upward.

CylonicRaider committed:
Move static field below nested classes

Hmm... Perhaps we should just give up and move the nested classes to the
bottoms of their parents.

CylonicRaider committed:
Editorial comment fixes

N00bySumairu commented:
The session cookie is stored at the level of the ConnectionManager though, and ConnectionEvents only bubble up to ConnectionImpl. Creating a new EventListener setup just for session cookie management honestly seems somewhat overkill to me, although it would probably be the solution closest to how data / events currently flow within the app. So I'd rather go with a new listener.

CylonicRaider commented:
Hmm... I'd argue that good design outranks implementation ease, and stuffing an unrelated method into `Connection` feels poor.

As far as implementation ease goes, note that one does not even need to introduce the method into `Connection` at least, as the method's intended user `EuphoriaWebSocketClient` is explicitly bound to a `ConnectionImpl` anyway.

CylonicRaider commented:
...The same consideration as the second paragraph also applies to `ConnectionManager`.

N00bySumairu commented:
So how do you want this to work?

- Have ConnctionManager listen to SessionCookieUpdateEvents by registering itself with EuphoriaWebSocketClients (which currently wouldn't work, as it doesn't have direct access to those.
- Have EuphoriaWebSocketClients call some onSessionCookieUpdate method on `parent.parent`, which is also not very pretty.
- Modify Connection after all to somehow forward session cookie updates

CylonicRaider commented:
Hmm... Another option would be to declare session cookies an implementation detail and drop the methods from the interfaces entirely. (It is actually a bit *less* coding effort than the PR's code, although a lot of arguing effort has already sunk into this particular hole.)

However, we need to consider the overall design goal of the connection API (at least until we burn and re-make it): To be an interface that provides no-fuss access to connections to Euphoria rooms. From a user's perspective, having no control over the connection API magically persisting session cookies appears highly unsatisfactory; it would be much better if it provided API surface to manage that.

Therefore, I propose adding a `SessionCookieStore` interface with appropriate get/put methods (also parametrized with, say, a room name), an instance whereof is defined to be conveyed to the `ConnectionManager`, from where it can somehow find its way to `EuphoriaWebSocketClient`, which can use it directly.

If we look closely (and in particular keep the room name parametrization in mind), that `SessionCookieStore` looks awfully close to a `Map<String, HttpCookie>`, so we might not even need a new interface... although we could still add a getter of the cookie store to `ConnectionManager`.

N00bySumairu commented:
Making cookies room based seems more like a maybe nice to have to me (which some might not even want; i.e. some might want one agent id across all rooms), the reference web client uses one cookie for all sessions regardless of the associated room after all, if I am not mistaken (unless you mess with private tabs, cookie deletion and tab containers). Making them room based also opens another rabbit hole to per-room cookie clearing, settings, etc. Is there demand for this? As far as I can tell, most would want one agent id for all rooms, across app re-launches, or ever-changing agent ids.

CylonicRaider commented:
Well... The *implementation* need not consider the room name, but... it sounds like something that *should* be in the API.

N00bySumairu committed:
Redo session cookie storage

Introduce a SessionCookieStore interface and refactor code to work with
it.

CylonicRaider commented:
- I ssspot misssing blank linesss againsss...
- The order of the session cookie putting and getting [sic!] helpers seems reversed.
- The blank lines in `SessionCookieStoreImpl` are unbalanced, ow!
- It might be worthwhile to add a comment that we are deliberately ignoring room names in the same class.

CylonicRaider committed:
Adjust spacing, comment line wrapping, and imports

CylonicRaider committed:
Tweak session cookie implementation details

CylonicRaider commented on app/src/main/java/io/euphoria/xkcd/app/impl/connection/ConnectionManagerImpl.java:
```diff
@@ -35,7 +34,7 @@ public synchronized Connection getConnection(String roomName) {
         ConnectionImpl conn = connections.get(roomName);
         if (conn == null) {
-            conn = new ConnectionImpl(this, roomName, settings.shouldContinuePrevSession() ? settings.getSessionCookie() : null);
+            conn = new ConnectionImpl(this, roomName);
```
Looks like the `shouldCountinuePrevSession` check went under during the refactoring; I'll re-add it before merging.

CylonicRaider replied:
Resolved in 8ffdc1c84a6c2dcee6f8336333a343def58ffa09.

CylonicRaider committed:
Revive session resumption check

CylonicRaider committed:
Editorial Javadoc improvements

CylonicRaider committed:
More minor whitespace and import adjustments
